File: dtmf.ts

namespace zkm {

    export enum DtmfPin {
        P0 = AnalogPin.P0,
        P1 = AnalogPin.P1,
        P2 = AnalogPin.P2,
    }

   /**
    */
    //% blockId=zkm_start_dtmf block="start process from pin %dtmfPin1 | and %dtmfPin2"
    export function startProcess(dtmfPin1: DtmfPin, dtmfPin2: DtmfPin) {
        startDtmfService(dtmfPin1, dtmfPin2)
    }

    //% shim=zkm::startDtmfService
    export function startDtmfService(dtmfPin1: number, dtmfPin2: number): void { return }

    /**
     * todo
     */
    //% blockId="zkm_playTone" block="playtone  %tone | for %duration | ms"
    export function play(tone: DtmfTone, duration: number): void {
        return playTone(tone, duration);
    }

    //% shim=zkm::playTone
    export function playTone(tone: DtmfTone, duration: number): void { }

    /**
    * @tone the tone, eg:DtmfTone.Tone_1
    * @handler action
    */
    //% weight=70
    //% blockId=zkm_on_tone block="on detect tone %tone"
    export function onToneDetected(tone: DtmfTone, handler: Action) {
        onTone(tone, handler)
    }

    //% shim=zkm::onTone
    export function onTone(tone: DtmfTone, handler: Action): void { return }

}

File: dtmf.cpp
#include "pxt.h"
#include "dtmfService.h"
#include "Ticker.h"
#include "dtmf.h"

using namespace pxt;

namespace zkm
{

static DtmfService *_dtmfService = NULL;
static Ticker dtmfTimer;
static Action _handler[DTMF_TONE_COUNT];

TwoTone DTMFTONES[DTMF_TONE_COUNT] = {
    {FREQ_L1, FREQ_H1}, // DtmfTone::Tone_1
    {FREQ_L1, FREQ_H2}, // DtmfTone::Tone_2
    {FREQ_L1, FREQ_H3}, // DtmfTone::Tone_3
    {FREQ_L2, FREQ_H1}, // DtmfTone::Tone_4
    {FREQ_L2, FREQ_H2}, // DtmfTone::Tone_5
    {FREQ_L2, FREQ_H3}, // DtmfTone::Tone_6
    {FREQ_L3, FREQ_H1}, // DtmfTone::Tone_7
    {FREQ_L3, FREQ_H2}, // DtmfTone::Tone_8
    {FREQ_L3, FREQ_H3}, // DtmfTone::Tone_9
    {FREQ_H4, FREQ_H2}, // DtmfTone::Tone_0
    {FREQ_L1, FREQ_H4}, // DtmfTone::Tone_A
    {FREQ_L2, FREQ_H4}, // DtmfTone::Tone_B
    {FREQ_L3, FREQ_H4}, // DtmfTone::Tone_C
    {FREQ_L4, FREQ_H4}, // DtmfTone::Tone_D
    {FREQ_H4, FREQ_H1}, // DtmfTone::Tone_STAR
    {FREQ_H4, FREQ_H3}, // DtmfTone::Tone_HASH
};

void playTones()
{
    _dtmfService->process();
}

//%
void startDtmfService(int dtmfPin1, int dtmfPin2)
{
    if (NULL != _dtmfService)
    {
        return;
    }

    MicroBitPin *pin1 = getPin(dtmfPin1);
    if (!pin1)
    {
        return;
    }
    MicroBitPin *pin2 = getPin(dtmfPin2);
    if (!pin2)
    {
        return;
    }

    _dtmfService = new DtmfService(pin1->name, pin2->name);

    dtmfTimer.attach_us(&playTones, 5 * 10000);
}

//%
void playTone(DtmfTone tone, int duration)
{
    if (NULL == _dtmfService)
    {
        return;
    }
    TwoTone twoTone = DTMFTONES[tone];

    _dtmfService->playTone(twoTone, duration);
}

//%
void onTone(DtmfTone tone, Action handler)
{
    _handler[tone] = handler;
    pxt::incr(_handler[tone]);
}


void notifyToneDetected(DtmfTone tone)
{
    Action handler = _handler[tone];
    if (handler)
    {
        pxt::runAction0(handler);
    }
}

}